biochatter/vectorstore_host.py

```diff
@@ -1,196 +1,409 @@
+    def store_embeddings(self, documents: List[Document]) -> str:
+        """
+        Store documents to database.
+        Arguments;
+            documents List[Documents]: documents array, usually from DocumentReader.load_document
+                                       DocumentReader.document_from_pdf, DocumentReader.document_from_txt
+        Returns:
+            str: document id
+        """
+        if len(documents) == 0:
             return
+        return self._insert_data(documents)
+    def _build_embedding_search_expression(self, meta_ids: List[Dict]) -> Optional[str]:
+        """
+        Build search expression for embedding collection. The generated expression is like: "meta_id in [{id1}, {id2}, ...]
+        Parameters:
+            meta_ids: the array of metadata id in metadata collection
+        Returns:
+            str: search expression or None
+        """
+        if len(meta_ids) == 0:
+            return None
+        built_expr = '''meta_id in ['''
+        for item in meta_ids:
+            id = f'"{item["id"]}",'
+            built_expr += id
+        built_expr = built_expr[:-1]
+        built_expr += ''']'''
+        return built_expr
+    
+    def _join_embedding_and_metadata_results(self, result_embedding: List[Document], result_meta: List[Dict]) -> List[Document]:
+        """
+        Join the search results of embedding collection and results of metadata.
+        Parameters:
+            result_embedding List[Document]: search result of embedding collection
+            result_meta List[Dict]: search result of metadata collection
+        Returns:
+            List[Document]: combined results like [{page_content: str, metadata: {...}}]
+        """
+        def _find_metadata_by_id(metadata: List[Dict], id: str) -> Optional[Dict]:
+            for d in metadata:
+                if str(d["id"]) == id:
+                    return d
+            return None
+
+        joined_docs = []
+        for res in result_embedding:
+            found = _find_metadata_by_id(result_meta, res.metadata["meta_id"])
+            if found is None: # discard
+                logger.error(f"Failed to join meta_id {res.metadata['meta_id']}")
+                continue
+            joined_docs.append(
+                Document(page_content=res.page_content, metadata=found)
+            )
+        return joined_docs
+    
+    def similarity_search(
+        self, query: str, k: int=3
+    ) -> List[Document]:
+        """
+        Similarity search according to query
+        This method will:
+        1). get all non-deleted meta_id and build to search expression for embedding collection
+        2). do similarity search in embedding collection
+        3). combine metadata and embeddings
+        Parameters:
+            query str: query string
+            k: the number of result
+        Returns:
+            List[Document]: search results
+        """
+        result_metadata = self._col_metadata.query(
```

Comment:
> @fengsh27 is there a way to actually delete embeddings in the milvus system? if deleted entities are always just denoted as "isDeleted", i would imagine that the database will grow to unnecessarily large dimensions after a while. seems suboptimal for the 'long-running DB' use case.

Good question. I chose to use an "isDeleted" flag rather than hard deleting for the reason that it allows the possibility to "undo" a deletion by simply flipping the flag back to false. This can be handy for recovering accidentally deleted data. 

For the large dimensions issue, I think we can introduce a new field named "deleteDate." This would enable us to regularly remove items marked as "isDeleted," specifically those with a "deleteDate" occurring before a designated threshold. 

What's your opinion?

Reply:
That sounds like a good idea, I like the undo option. Doing a periodic cleanup could be fine, or even a manual one. Something like `.delete_unused()`.